Allow to serialize extension values as DER bytes strings

felixfontein opened the pull request:
This is another try to implement #5002 (Support for getting ASN.1 value for parsed extensions). 

It uses the existing backend function `_create_x509_extension()` to convert a extension created for this extension value to an OpenSSL extension, and then uses the same code as in https://github.com/pyca/cryptography/blob/70ccc2542cf3bf374861b55f6527424307722763/src/cryptography/hazmat/backends/openssl/decode_asn1.py#L239-L242 to extract the extension's value. As a handler map it uses the union of all handler maps in `encode_asn1.py` (I checked that the same OID in different mappings are mapped to the same function, in case someone wonders).

If the general approach looks good, I can add documentation and tests.

reaperhulk commented:
We're going to convert X509 encoding to Rust next so I'd prefer to tackle this after we do that conversion. With the rust code we can just call `asn1::write_single` to serialize to DER so implementing this would be very simple.

felixfontein commented:
@reaperhulk @alex I've reworked this PR to use the Rust code to serialize the extension values (and falls back to Python+OpenSSL for the ones not yet supported by the Rust code). Does the approach look sensible / acceptable? (I've added in the comments what needs to be updated once all extensions are serialized by Rust code.)

reaperhulk commented:
This PR took advantage of a window where we had exposed public methods for extension encoding. Those are now gone as we completed our x509 conversion. However, we can add methods back, we should just scope them to doing exactly what this method requires. It seems like what you really want is a single rust method `encode_extension` that takes an `ExtensionType` and returns the encoded bytes without regard to whether they're a CSR, cert, CRL, CRL entry, or OCSP extension?

I also think the name here should probably be `public_bytes` (which matches our DER serialization on `Name` as well).

felixfontein commented:
@reaperhulk thanks for the feedback! I've renamed it to `public_bytes` as you suggested, and added a rust function `x509::common::encode_extension_value` (its code inspired by `encode_extensions` above it), and it seems to work fine. This is the first Rust code I've written in some years now, so hopefully it is not too horrible :)

The failing CI check seems to be unrelated to this PR (or there is something very strange going on I don't understand...).

reaperhulk commented on src/rust/src/x509/common.rs:
```diff
@@ -646,6 +647,52 @@ pub(crate) fn encode_extensions<
+    )
+    .unwrap();
+
+    for encode_ext in [
```
I don't really like this approach but I'm curious what @alex thinks. I would prefer a new func that just goes through all extensions. However, you'd need to then have tests for serializing every ext to ensure coverage (which isn't necessarily a bad thing).

reaperhulk replied:
ping @alex

alex replied:
If we got rid of the thing where you can only serialize some extensions with some structures, then this would be much easier. Any objections to doing that?

reaperhulk replied:
So allow users to serialize whatever, even if it isn't valid for the structure based on our reading of RFC 5280? I have mixed feelings about that, but it is still syntactically valid ASN.1.

alex replied:
Does 5280 say you can't put a SAN in an OCSP response? I just thought that was us being practical.

reaperhulk replied:
5280 has no opinions about OCSP at all 😄 But 6960 says under section 4:

```
This section defines some standard extensions, based on the extension
   model employed in X.509 version 3 certificates (see [RFC5280]).
   Support for all extensions is optional for both clients and
   responders.  For each extension, the definition indicates its syntax,
   processing performed by the OCSP responder, and any extensions that
   are included in the corresponding response.
```

CSR, OCSP, and Cert/CRL are all in different RFCs, but in 5280 (where cert and CRL are) there are distinct sections for cert and CRL extensions.

All that said, since arbitrary extensions are part of the spec an extension that just so happens to be exactly the same as one defined for another type is valid so it's not a big stretch to say we should just collapse it all into one big parser.

felixfontein replied:
Ok, so how should we proceed with this? Is this something you want to implement? Or should I implement that (in this PR, in a separate PR, ...)?

reaperhulk replied:
Let's do this as two PRs:
* Unify the serialization into one function and call it from all the serialization methods.
* Use that method here

reaperhulk replied:
And if you're willing to supply that first PR that'd be great!

felixfontein replied:
I can definitely give it a try!

reaperhulk reviewed (commented):
also have a changelog conflict

felixfontein commented on docs/x509/reference.rst:
```diff
@@ -1589,6 +1589,14 @@ X.509 Extensions
 
         Returns the OID associated with the given extension type.
 
+    .. method:: public_bytes
```
Thanks, fixed in e1dc35f767613ca1a5ac505302b40b23a5c6a954.

felixfontein commented on src/cryptography/x509/extensions.py:
```diff
@@ -90,6 +91,9 @@ def __init__(self, msg: str, oid: ObjectIdentifier) -> None:
 class ExtensionType(metaclass=abc.ABCMeta):
     oid: typing.ClassVar[ObjectIdentifier]
 
+    def public_bytes(self) -> bytes:
```
I guess there are some choices:
1. Make this abstract, and implement it (the same way) in every class derived from `ExtensionType`. That's a lot of copy'n'paste, though.
2. Make this abstract, add an internal intermediate class from which the other supplied extensions are derived which implements this.
3. Convert `public_bytes` into a free function, then users have to call `x509.extension_type_public_bytes(my_ext)` instead.

An advantage of 1. and 2. would be that user-defined extension classes are forced to implement this function by themselves. Besides that, I don't really like them (and 3.) much more than the current implementation.

reaperhulk replied:
I'm leaning towards option 1. It's verbose, but I'm against adding inheritance trees (option 2). Free func (option 3) would be fine, except it's not consistent with the way we do things in our APIs in general.

felixfontein replied:
Hmm, should the method in `ExtensionType` be abstract, or contain a `raise NotImplementedError(...)`? If it's abstract, this potentially breaks existing code which already inherits from `ExtensionType`.

reaperhulk replied:
Curse mypy for making us inherit from abcs we previously just registered... We haven't historically worried about practical breakage in expanding our interface requirements, but maybe we need to this time. If we do then we should raise `NotImplementedError` I suppose. @alex probably has an opinion here.

felixfontein replied:
I've picked an abstract method in b43cf7ad920d6d11f64c4eda58c93415931077c4, but it's trivial to change.

felixfontein replied:
... aaand I changed it in ea13b8de25eeca78dd4a588f761396eceac4c26b since there were some tests failing which defined their own extension types. I can revert this new commit and fix the tests, but IMO it's probably better to go with `NotImplementedError`...

felixfontein replied:
Interestingly the result of this approach is that I found out that there's no code that serializes `SignedCertificateTimestamps` right now. Since this extension can appear in OCSP responses (that's where I was able to get one from in 257172e5441bb6b1edcf8872b003a83c7a35f0b3), this looks like a potential oversight to me? CC @alex

reaperhulk replied:
We didn't previously support this because OpenSSL has no APIs for it. If we want to add support for this let's just treat that as a separate issue.

felixfontein replied:
I've created #6580 for `SignedCertificateTimestamps`.

felixfontein commented:
Rebased on top of #6571.

felixfontein commented:
Rebased after #6571 was merged, and CI is green on the first try :tada:

reaperhulk reviewed (approved):
approving even though our serialization for KU needs to properly account for unused bits because this code will improve testing for the fix.

felixfontein commented:
@reaperhulk @alex thanks a lot for reviewing and merging!
